app/Console/Commands/AutoUpsertHistories.php

```diff
@@ -64,19 +64,14 @@ public function handle(): void
          * - updated_at
          */
         $historiesPerCycle = intdiv(65_000, 16);
-
         $key = config('cache.prefix').':histories:batch';
         $historyCount = Redis::connection('announce')->command('LLEN', [$key]);
 
-        for ($historiesLeft = $historyCount; $historiesLeft > 0; $historiesLeft -= $historiesPerCycle) {
-            $histories = Redis::connection('announce')->command('LPOP', [$key, $historiesPerCycle]);
-
-            if ($histories === false) {
-                break;
-            }
-
-            $histories = array_map('unserialize', $histories);
+        $allHistories = Redis::connection('announce')->command('LPOP', [$key, $historyCount]);
+        $allHistories = array_map('unserialize', $allHistories);
+        $chunks = array_chunk($allHistories, $historiesPerCycle);
 
+        foreach ($chunks as $histories) {
```

Comment:
It's better to keep redis queries short and simple as it's single-threaded and can not handle requests concurrently. It should also be compared to false, otherwise it causes errors if no user announces between intervals.